### PR TITLE
Refactor the Debugger detect_stalled_training_job_and_stop.ipynb notebook

### PR DESCRIPTION
**Pull request by mchoi8739**
*Issue #, if available:
https://github.com/aws/amazon-sagemaker-examples/issues/1584

*Description of changes:
Refactor the Debugger [detect_stalled_training_job_and_stop.ipynb](https://github.com/aws/amazon-sagemaker-examples/blob/master/sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb) notebook.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb [hunk 1]
```diff
@@ -4,30 +4,22 @@
    "cell_type": "markdown",
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>What if customer is not using default container?</p><p>We should have a section explaining that customers can do hook.save_scalar periodically in their training/validation loop as heartbeat mechanism.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>Rule will always emit cloudwatch event even in case when rule is not sure about which training job to stop. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>For DLC you don't need to provide Collections to save.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>Where can user see rule job logs? </p><p> </p><p>FAQ: Why is my training not stopping? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>I don't think the StalledTrainingRule has a default tensor collection to watch, so I specified one as an example. It has to have a tensor collection to watch (based on the <a href="https://github.com/awslabs/sagemaker-debugger-rulesconfig/blob/master/smdebug_rulesconfig/rule_config_jsons/ruleConfigs.json" target="_blank">ruleConfigs.json</a> file). For DLC, based on my understanding, it is required to specify a collection either throught the <code>Rule.sagemaker</code>'s collections_to_save or the <code>DebugHookConfig()</code> API, unless the rule is pre-configured as provided in the <a href="https://github.com/awslabs/sagemaker-debugger-rulesconfig/blob/master/smdebug_rulesconfig/rule_config_jsons/ruleConfigs.json" target="_blank">ruleConfigs.json</a> file.</p>

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>This new script outputs the rule job status every 15 minutes as the following:</p><pre class="ql-syntax" spellcheck="false">Current job status: [PrimaryStatus: Stopped, SecondaryStatus: Stopped] | StalledTrainingRule Rule Evaluation Status: IssuesFound
</pre><p>I created this output script to match the descriptions, and customers can easily read it.</p><p> </p><p> </p><p>The previous script for CW, <code>get_rule_jobs_cw_urls(estimator)</code>, outputs:</p><pre class="ql-syntax" spellcheck="false">{'StalledTrainingRule': 'https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logStream:group=/aws/sagemaker/ProcessingJobs;prefix=smdebug-stalled-demo-16019-StalledTrainingRule-f8bac2de;streamFilter=typeLogStreamPrefix'}
</pre><p>It is unclear what to do with this url. I can add a statement to check CW if they want to look far a detailed rule job logs.</p>

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: > I don't think the StalledTrainingRule has a default tensor collection to watch, so I specified one as an example. Each rule has to have a tensor collection to watch (based on the [ruleConfigs.json](https://github.com/awslabs/sagemaker-debugger-rulesconfig/blob/master/smdebug_rulesconfig/rule_config_jsons/ruleConfigs.json) file, some have default collections, some don't). For DLC, based on my understanding, it is required to specify a collection either through the `Rule.sagemaker`'s `collections_to_save` or the `DebugHookConfig()` API, unless the rule is pre-configured as provided in the [ruleConfigs.json](https://github.com/awslabs/sagemaker-debugger-rulesconfig/blob/master/smdebug_rulesconfig/rule_config_jsons/ruleConfigs.json) file.

Stalled training rule checks if there is 'any' tensor emitted. By default in DLC, we have loss emitted every 500 steps, so it should just work. 
When you don't have DLC, you would need to emit tensor on your own.

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>For example, vanishing_gradient() or loss_not_decreasing() rules are having <code>gradients</code>  and <code>losses</code>  tensor collections by default in their .json config file, so we don't have to specify any tensor collection. But the ones without the default collections, I think it needs to be specified</p>

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: We should provide detailed steps on how to view CW logs for a rule.  Actually it should be in sagemaker doc page doc.

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>I recently added this page, and provided instruction to check the rule job logs: https://docs.aws.amazon.com/sagemaker/latest/dg/debugger-cloudwatch-lambda.html</p>

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Ok, can you try running the notebook after removing collection to save?

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: https://github.com/aws/amazon-sagemaker-examples/pull/1592#discussion_r499962051
If this is true, I would say saving `losses` tensors using DebugHookConfig() should conflict. But that never happened.

For example, if I include `vanishing_gradient()` rule and add DebuggerHookConfig(collection_configs="gradients"), then it returns a confliction error, saying that the duplicated collections names are used.

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: This doc doesn't tell how to get to actual CW log. We should write exact steps in the docs.
For example, customer go to training job, he sees 1000s of training job. 
He goes to processing job he sees 1000s of processing job. How should he go ahead and find the processing job that he is interested in. 
How does he get to logs from ProcessingJob page?

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Does Debugger collect tensors with regex ".*loss" by default, or the "losses" collection by default?

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: > Ok, can you try running the notebook after removing collection to save?
Yes, I can run it without the collection to save, but the training job has the 10 minute sleep time, so it will not change the result tho?

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: 2 minute threshold would fire the rule, if there is 10 minute sleep in training.

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: I see, good point. I will include it to the doc and the notebook again.

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Yep that's what I meant. The training job ran and slept well to stop after 2 mins, and it will be the same for with and without the collection_to_save option. Then do you think we need to remove it to simplify the rule config?

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: yes, if users are using DLC they don't  need to add collections to save for this rule. Though we can make it clear, why? Saying that smdebug default saves losses every 500 steps, which will act as heartbeat for training job.

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Added the hook registration code lines as comments in the training script. Also mentioned about it in the notebook.

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Added a statement about this

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Incorporated

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Incorporated

Comment (Vikas-kum) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: I guess  it did worked when you removed collection configuration, is that right?

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Yes, it did.

File: sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb [hunk 2]
```diff
@@ -4,30 +4,24 @@
...
```
Comment (Vikas-kum) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>nit: manual hook registration on / manual hook registration on.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Review by Vikas-kum — APPROVED** [review]
Looks good. Small nit and a confirmation question.
Should be good to go once addressed.

**Review by Vikas-kum — DISMISSED** [review]

File: sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb (hunk 2 above)
Comment (TEChopra1000) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p><strong>1) </strong></p><p>"This notebook guides you how to use the <code>StalledTrainingRule</code> built-in rule." needs to be reworded.</p><p> </p><p>Example:</p><p>"This notebook shows you how to use the <code>StalledTrainingRule</code> built-in rule."</p><p> </p><p><strong>2) </strong></p><p>In this sentence "This functionality helps you monitor the training job status and save redundant resource usage." did you mean to say "This functionality helps you monitor the training job status and <u>reduces</u> redundant resource usage."</p><p> </p><p><strong>3)</strong></p><p>The sentence: "If you want to explore using the manual hook registration, " --&gt; if you want to explore <em>what</em>? Maybe "If you want to use debugger with..."?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (TEChopra1000) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>"The unique prefix" --&gt; "A unique prefix".</p><p> </p><p>" it will fallback to safe mode and not take action .. " --&gt; " it falls back to safe mode and does not stop the training job."</p><p> </p><p>"Such rule evaluation process goes on in parallel while the training jobs are running" --&gt; can you replace "Such" with something else, maybe "This"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (TEChopra1000) [inline sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: <p>"trining" --&gt; "training"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/1592/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Feedback incorporated

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Feedback incorporated

Comment (mchoi8739) [reply sagemaker-debugger/tensorflow_action_on_rule/detect_stalled_training_job_and_stop.ipynb:4]: Feedback incorporated

**Review by TEChopra1000 — APPROVED** [review]
